Kafka Node Pools: KRaft (ZooKeeper-less Apache Kafka) (#382)
strimzi/strimzi.github.io: _posts/2023-08-29-kafka-node-pools-kraft.md

The Kafka nodes in the KRaft mode have _controller_ and/or _broker_ roles.
The nodes with the _controller_ role are responsible for maintaining the Kafka metadata, handling leader elections, and helping to bootstrap the Kafka cluster.
They assume the responsibilities previously handled by the ZooKeeper nodes.
The nodes with the _broker_ role have the same responsibilities as the Kafka brokers in the ZooKeeper-based Kafka cluster.
They receive the messages from the producers, store them, and pass them to the consumers.
One node can also have both roles - a controller and a broker -  at the same time.
This offers a number of configuration options for your KRaft-based Kafka cluster.

You might start with a small Kafka cluster consisting only of combined nodes that all have controller and broker roles.
Typically, it would have three nodes.
But if needed you can also run _a whole Kafka cluster_ in a single node.
Such a cluster is suitable for example for development and testing.
Or - in some cases - it might be even considered for a small production cluster.

![Three-node Kafka cluster with combined nodes](/assets/images/posts/2023-08-29-kafka-node-pools-kraft-combined-nodes.png)

> For a big production cluster, you would normally use dedicated controller nodes and dedicated broker nodes.

Author: I'm open to suggestions on how to make this less _controversial_ if you feel that is how it looks. What is the thing which is the problem? That the combined nodes `might be considered for a small production cluster`? Or the `big production cluster` formulation? The goal is not to provide definitive guidance on when to use what. So I think that as long as the blog post sufficiently explains why are all these modes important, it does not need to be controversial.

Would this work for you? (***bold*** parts are new/changed, ~~strikethrough~~ will be removed)

----

You might start with a small Kafka cluster consisting only of combined nodes that all have controller and broker roles.
Typically, it would have three nodes.
But if needed you can also run _a whole Kafka cluster_ in a single node.
Such a cluster is suitable for example for development and testing.
***It might be also useful in other resource-constraint environments.
For example when running Kafka at the edge.***
~~Or - in some cases - it might be even considered for a small production cluster.~~

![Three-node Kafka cluster with combined nodes](/assets/images/posts/2023-08-29-kafka-node-pools-kraft-combined-nodes.png)

***For a production cluster, you would normally use dedicated controller nodes and dedicated broker nodes.***
You would typically maintain three controller nodes for availability, and add as many broker nodes as you need.
***This is also the architecture to which you would migrate from a ZooKeeper-based cluster.***

![Three-node Kafka cluster with dedicated controller and broker nodes](/assets/images/posts/2023-08-29-kafka-node-pools-kraft-separate-nodes.png)

And somewhere in between is an architecture where some nodes have both controller and broker roles and some have only the broker role.
~~It is a good fit for example when you run on bare metal and have only powerful worker nodes that you want to be fully utilized and that would be too big to run only as controllers.~~
It is also an intermediate step when migrating between the previous two architectures. 

----

That said, my feeling is that the examples of when it might be used I use there are realistic, in line with the expectations of many users, and correspond to how users will expect to use them. Blindly insisting on dedicated controller nodes for every production cluster will be very shortsighted and will not do any good for Kafka. But no problem to leave this out of this blog post.

> For a big production cluster, you would normally use dedicated controller nodes and dedicated broker nodes.

Reviewer: Thanks, I like your suggested construction much better! @mimaison @showuon are you also happy with this?

> For a big production cluster, you would normally use dedicated controller nodes and dedicated broker nodes.

Reviewer 2: Looks good. Thanks.

> For a big production cluster, you would normally use dedicated controller nodes and dedicated broker nodes.

Author: Ok, I updated the PR @tombentley

You would typically maintain three controller nodes for availability, and add as many broker nodes as you need.

![Three-node Kafka cluster with dedicated controller and broker nodes](/assets/images/posts/2023-08-29-kafka-node-pools-kraft-separate-nodes.png)

And somewhere in between is an architecture where some nodes have both controller and broker roles and some have only the broker role.
It is a good fit for example when you run on bare metal and have only powerful worker nodes that you want to be fully utilized and that would be too big to run only as controllers.
It is also an intermediate step when migrating between the previous two architectures. 

![Three-node Kafka cluster with combined nodes](/assets/images/posts/2023-08-29-kafka-node-pools-kraft-semi-combined-nodes.png)

Our users employ Strimzi for various scenarios, spanning from small development and test clusters to large-scale multi-node production clusters.
So we want to support all of these architectures.
We also want to make it possible to transition between them.

> But how do we model the different architectures?

Reviewer: ```suggestion
But how do we model the different architectures as Kubernetes resources?
```

And somewhere in between is an architecture where some nodes have both controller and broker roles and some have only the broker role.
It is a good fit for example when you run on bare metal and have only beefy worker nodes that you want to be fully utilized and that would be too big to run only the controller node.
It is also an intermediate step when migrating between the previous two architectures. 

![Three-node Kafka cluster with combined nodes](/assets/images/posts/2023-08-29-kafka-node-pools-kraft-semi-combined-nodes.png)

Our users use Strimzi for various use cases from development and test clusters up to large production clusters with many nodes.
So we want to support all of these architectures.
We also want to make it possible to transition between them.
But how do we model the different architectures?
The answer is - using the `KafkaNodePool` resources.

### Node pools and KRaft

KRaft and its various architectures were one of the main motivations when designing node pools.
That is also why KRaft and its roles are directly part of it.
Each `KafkaNodePool` resource has a mandatory field called `roles` in its `spec` section.

> We already talked about it already in the [introduction](https://strimzi.io/blog/2023/08/14/kafka-node-pools-introduction/) blog post.

Reviewer 3: ```suggestion
We mentioned this in the blog post [introducing node pools](https://strimzi.io/blog/2023/08/14/kafka-node-pools-introduction/).
```

The `roles` field contains a list of roles that the nodes from the node pool should take.

> In ZooKeeper-based clusters, the only available role is the `broker` role.

Reviewer 3: ```suggestion
In ZooKeeper-based clusters, you should only configure the `broker` role.
```

> In KRaft clusters, you can choose between `controller` and/or `broker`.

Reviewer 4: ```suggestion
In KRaft-based clusters, you can configure `controller` and/or `broker`.
```

And Strimzi will configure the Kafka nodes belonging to the given pool accordingly.
There always has to be at least one node pool that has the `controller` role and at least one node pool with the `broker` role.
It can be the same node pool for both roles or it can be in a different node pool.
But both roles have to be present to have a working Kafka cluster.
And you can use the roles to configure the various architectures.

So for example, to create the KRaft cluster with combined nodes, you can create one `KafkaNodePool` and configure it to use both roles - `controller` as well as `broker`.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaNodePool
metadata:
  name: combined
  labels:
    strimzi.io/cluster: my-cluster
spec:
  replicas: 3
  roles:
    - controller
    - broker
  # ...
```

> If you want to have a Kafka cluster with a dedicated `controller` and `broker` nodes, you can create two `KafkaNodePool` resources.

Reviewer 3: ```suggestion
If you want to have a Kafka cluster with dedicated `controller` and `broker` nodes, you can create two `KafkaNodePool` resources.
```

One with the `controller` role.

> And other with the `broker` node.

Reviewer 3: ```suggestion
And another with the `broker` node.
```

  labels:
    strimzi.io/cluster: my-cluster
spec:
  replicas: 6
  roles:
    - broker
  # ...
```

Finally, if you want to use some combined nodes and some broker-only nodes, you can configure the Kafka cluster that way as well.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaNodePool
metadata:
  name: combined
  labels:
    strimzi.io/cluster: my-cluster
spec:
  replicas: 3
  roles:
    - controller
    - broker
  # ...
---

apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaNodePool
metadata:
  name: broker
  labels:
    strimzi.io/cluster: my-cluster
spec:
  replicas: 6
  roles:
    - broker
  # ...
```

You can find full examples of the custom resources on our [GitHub](https://github.com/strimzi/strimzi-kafka-operator/tree/main/examples/kafka/nodepools).
If you want to try it, do not forget that you have to enable the [`UseKRaft` feature gate](https://strimzi.io/docs/operators/0.36.1/full/deploying.html#ref-operator-use-kraft-feature-gate-str) as well.

### Current limitations

While the work on the KRaft support is progressing, it is still not completely ready.
At the time of writing (Strimzi 0.36 / Apache Kafka 3.5.1), there are still some important features missing.

> Right now, you can use Strimzi to deploy a Kafka cluster will all of the architectures mentioned earlier.

Reviewer 3: ```suggestion
Right now, you can use Strimzi to deploy a Kafka cluster with all of the architectures mentioned in this post.
```

But Strimzi currently does not support rolling updates of the nodes that have the controller role only.
It also does not support the migration from ZooKeeper to KRaft or upgrades of KRaft clusters.
A full list of the current Strimzi KRaft limitations can be found in our [documentation](https://strimzi.io/docs/operators/latest/full/deploying.html#ref-operator-use-kraft-feature-gate-str).

> There are also still many limitations in Apache Kafka itself.

Reviewer 3: ```suggestion
There are also still many limitations to KRaft in Apache Kafka itself.
```

For example missing support for JBOD storage or Admin API support for managing the controller nodes.

> So, if you want, give this a try.
> But be aware the missing features.

Reviewer 3: ```suggestion
So, feel free to give the KRaft-based node pool setups a try, but do keep in mind the existing limitations.
```

### Conclusion

> KRaft might not be production ready yet.
> But it's clear, that it is the future of Apache Kafka, and supporting the various KRaft architectures is important for Strimzi.
> And the Kafka node pools play a very important role in it.

Reviewer 3: ```suggestion
KRaft might not be production-ready yet, but it is the future of cluster management in Apache Kafka.
As such, Strimzi is dedicated to supporting the various KRaft architectures, with Kafka node pools playing a very important role in this commitment.
```
